Add doc for using custom plugins and JAR files (#7304)
elastic/cloud-on-k8s: docs/orchestrating-elastic-stack-applications/logstash.asciidoc

      - name: download-postgres
        command: ["/bin/sh"]
        args: ["-c", "curl -o /data/postgresql.jar -L https://jdbc.postgresql.org/download/postgresql-42.6.0.jar"]
        volumeMounts:
          - name: workdir
            mountPath: /data
      containers:
        - name: logstash
          volumeMounts:
            - name: workdir
              mountPath: /usr/share/logstash/jars <1>
  pipelines:
    - pipeline.id: main
      config.string: |
        input {
          jdbc {
             jdbc_driver_library => "/usr/share/logstash/jars/postgresql.jar" <1>
             jdbc_driver_class => "org.postgresql.Driver"
                :
                :
          }
        }
----
<1> Referring to the external file should match the `mountPath` of the `container`

=== Adding files using a custom image

This example downloads the same `postgres` JDBC driver, and adds it to the {ls} classpath in the Docker image.

First, create a Dockerfile based on the {ls} Docker image, and download the JDBC driver, storing it alongside the other JAR files in the {ls} classpath:

[source,subs="attributes,+macros,callouts"]
----
FROM docker.elastic.co/logstash/logstash:{version}

> RUN curl -o /usr/share/logstash/logstash-core/lib/jars/postgresql.jar -L https://jdbc.postgresql.org/download/postgresql-42.6.0.jar <1>

Reviewer: This is not working as expected.  The URL is interpreted as a hyperlink (href).  

<img width="709" alt="Screen Shot 2024-01-17 at 4 47 38 PM" src="https://github.com/elastic/cloud-on-k8s/assets/35154725/c6ab3ed6-efa1-44ff-af47-e5a34c23a8f5">

I can look for examples and/or play around with how to fix this.

Reviewer: I have it looking right with this `["source","sh",subs="attributes,callouts"]`, but CURL isn't working: 

<img width="708" alt="Screen Shot 2024-01-17 at 5 08 28 PM" src="https://github.com/elastic/cloud-on-k8s/assets/35154725/8e084bad-3712-446a-a1d2-d4ef510c1d8a">

I'll keep looking.

> First, create a Dockerfile based on the {ls} Docker image, and download the JDBC driver, storing it alongside the other JAR files in the {ls} classpath:
> 
> [source,subs="attributes,+macros,callouts"]
> ----
> FROM docker.elastic.co/logstash/logstash:{version}
> 
> RUN curl -o /usr/share/logstash/logstash-core/lib/jars/postgresql.jar -L https://jdbc.postgresql.org/download/postgresql-42.6.0.jar <1>
> ----
> <1> Placing the JAR file in the `/usr/share/logstash/logstash-core/lib/jars` folder places it on Logstash's classpath.
> 
> Then after building and deploying the custom image (refer to <<{p}-custom-images>> for more details), include it in the {ls} manifest:

Reviewer: ```suggestion
First, create a Dockerfile based on the {ls} Docker image.
Download the JDBC driver, and save it alongside the other JAR files in the {ls} classpath:

From docker.elastic.co/logstash/logstash:{version}, run: 
[source,js]
----
curl -o /usr/share/logstash/logstash-core/lib/jars/postgresql.jar -L https://jdbc.postgresql.org/download/postgresql-42.6.0.jar <1>
----
<1> Placing the JAR file in the `/usr/share/logstash/logstash-core/lib/jars` folder adds it to the {ls} classpath.

After you build and deploy the custom image, include it in the {ls} manifest.
(Check out <<{p}-custom-images>> for more details.)
```

=== Adding files using an initContainer

This example creates an `initContainer` to download a PostgreSQL JDBC driver JAR file, and place it in a volume mount accessible to the main `container`, and then use it in a JDBC input in the pipeline configuration.

[source,yaml]
----
spec:
  podTemplate:
    spec:
      initContainers:
      - name: download-postgres
        command: ["/bin/sh"]
        args: ["-c", "curl -o /data/postgresql.jar -L https://jdbc.postgresql.org/download/postgresql-42.6.0.jar"]
        volumeMounts:
          - name: workdir
            mountPath: /data
      containers:
        - name: logstash
          volumeMounts:
            - name: workdir
              mountPath: /usr/share/logstash/jars <1>
  pipelines:
    - pipeline.id: main
      config.string: |
        input {
          jdbc {
             jdbc_driver_library => "/usr/share/logstash/jars/postgresql.jar"
             jdbc_driver_class => "org.postgresql.Driver"
             <2>
          }
        }
----
<1> Referring to the external file should match the `mountPath` of the `container`

> <2> Rest of plugin configuration goes here

Reviewer: ```suggestion
<2> Remainder of plugin configuration goes here
```

Reviewer: Maybe "remainder" instead of "rest" because "rest" can have multiple meanings.  
Or "Remaining plugin configuration goes here." 
You make the call.

First, create a Dockerfile based on the {ls} Docker image.
Download the JDBC driver, and save it alongside the other JAR files in the {ls} classpath:

From docker.elastic.co/logstash/logstash:{version}, run: 
[source,js]
----
curl -o /usr/share/logstash/logstash-core/lib/jars/postgresql.jar -L https://jdbc.postgresql.org/download/postgresql-42.6.0.jar <1>
----
<1> Placing the JAR file in the `/usr/share/logstash/logstash-core/lib/jars` folder adds it to the {ls} classpath.

After you build and deploy the custom image, include it in the {ls} manifest.
(Check out <<{p}-custom-images>> for more details.)

[source,yaml]
----
  count: 1
  version: {version} <1>
  image: <CUSTOM_IMAGE>
  pipelines:
    - pipeline.id: main
      config.string: |
        input {
          jdbc {
              <2>
             jdbc_driver_class => "org.postgresql.Driver"
              <3>
          }
        }

----
<1> Providing the correct version is always required as ECK reasons about APIs and capabilities available to it based on the version field.
<2> Note that when you place the JAR file on the {ls} classpath, you do not need to specify the `jdbc_driver_library` location in the plugin configuration.

> <3> Rest of plugin configuration goes here

Reviewer: ```suggestion
<3> Remainder of plugin configuration goes here
```

Reviewer: Your call.  Please go with whatever you decided for Line 972 here.

First, create a Dockerfile based on the logstash docker image, running `bin/logstash-plugin install` to install the appropriate plugin(s)

[subs="attributes,+macros,callouts"]
----
FROM docker.elastic.co/logstash/logstash:{version}

RUN bin/logstash-plugin install logstash-filter-tld
----

Then after building and deploying the custom image (see <<{p}-custom-images>> for more details), include it in the {ls} manifest:

[source,yaml]
----
spec:
  count: 1
  version: {version} <1>
  image: <CUSTOM_IMAGE>
----
<1> Providing the correct version is always required as ECK reasons about APIs and capabilities available to it based on the version field.

[id="({p}-logstash-jar-files"]
== Working with plugins that require additional files

Running {ls} may require additional files, such as JAR files needed to load JDBC drivers when using a JDBC or JMS plugin. To add these files, there are two options available - using an `initContainer` to add files before the main container start, or creating a custom docker image that includes the required files. See <<{p}-bundles-plugins>> for a run down of which option might be most suitable for you.

> === Using an initContainer

Reviewer: ```suggestion
=== Adding files using an initContainer
```

This example creates an `initContainer` to download a `postgres` JDBC Driver JAR file, and place it in a volume mount accessible to the main `container`, and then use it in a JDBC input in the pipeline configuration.

[source,yaml]
----
spec:
  podTemplate:
    spec:
      initContainers:
      - name: download-postgres
        command: ["/bin/sh"]
        args: ["-c", "curl -o /data/postgresql.jar -L https://jdbc.postgresql.org/download/postgresql-42.6.0.jar"]
        volumeMounts:
          - name: workdir
            mountPath: /data
      containers:
        - name: logstash
          volumeMounts:
            - name: workdir
              mountPath: /usr/share/logstash/jars <1>
  pipelines:
    - pipeline.id: main
      config.string: |
        input {
          jdbc {
             jdbc_driver_library => "/usr/share/logstash/jars/postgresql.jar" <1>
             jdbc_driver_class => "org.postgresql.Driver"
                :

>                 :

Author: To indicate that the rest of the config was snipped - @karenzone do we have a better way to indicate that content is snipped for brevity?

>                 :

Reviewer: We could introduce this as a _snippet_ to set expectations.  Then maybe have a blank section with a callout suggesting other stuff users might want to put there?  Something like this: 

<img width="715" alt="Screen Shot 2024-01-17 at 4 53 45 PM" src="https://github.com/elastic/cloud-on-k8s/assets/35154725/97d3a999-d335-450c-a93b-55a7d0b8c6cb">

>                 :

Author: Done

          }
        }
----
<1> Referring to the external file should match the `mountPath` of the `container`

> === Using a custom image

Reviewer: ```suggestion
=== Adding files using a custom image
```

Reviewer: For better context and SEO performance.

This example downloads the same `postgres` JDBC driver, and adds it to the Logstash classpath in the docker image

First, create a Dockerfile based on the logstash docker image, and download the JDBC driver, storing it alongside the other JAR files in the Logstash classpath:

> +

Reviewer 2: I think this is not needed. There is no ongoing enumeration or else to continue and it just renders as `+`
<img width="896" alt="Screenshot 2023-11-24 at 14 36 25" src="https://github.com/elastic/cloud-on-k8s/assets/697790/7dd4e6dc-ae82-490e-aa66-313ec68b2c16">

> [subs="attributes,+macros,callouts"]

Reviewer 2: ```suggestion
[source,subs="attributes,+macros,callouts"]
```
Maybe?

<1> Value of password to protect the Logstash keystore
<2> The syntax for referencing keys is identical to the syntax for environment variables

[id="{p}-logstash-custom-files-plugins"]

> == Adding plugins and files

Reviewer: Good content here.  There's not a lot of shared content here, and it seems like it would be easier to read and follow if you broke this topic into two topics:

* Adding a plugin
** Adding a plugin using an initContainer
** Adding a plugin using a custom image

* Adding more files
** Adding files using an initContainer
** Adding files using a custom image

Reviewer: The repetition in headings in intentional to provide the user with specific guideposts, and for better SEO.

=== Adding a plugin

To add plugins, the `bin/logstash-plugin` utility needs to be run to add the plugin to the logstash pod before it starts.

==== Using an initContainer

This example sets up an `initContainer` to install the `logstash-filter-tld` plugin:

[source,yaml]
----
spec:
  podTemplate:
    spec:
      initContainers:
      - name: install-plugins
        command: ["/bin/sh"]
        args: ["-c", "bin/logstash-plugin install logstash-filter-tld"]
----

The init container inherits:

* The image of the main container image, if one is not explicitly set.
* The volume mounts from the main container unless a volume mount with the same name and mount path is present in the init container definition
* The Pod name and IP address environment variables.

==== Using a custom image

> This example installs the same plugin, but uses a custom image, see XXX for more details

Reviewer 3: `, see XXX for more details` -> ~, see XXX the following~
